### DISCUSSION TITLE
How to create a project by importing it from another GitLab instance (it's private)

### DISCUSSION THREAD

**Discussion by w7089** [python-gitlab/python-gitlab#2524] [Q&A] [answered]
I'd like to create a new project by importing it from another GitLab instance using CLI (see screenshot from UI) where the source repository is private and requires authentication.

![image](https://user-images.githubusercontent.com/3884662/226729402-9d1908a9-8695-4e56-a6d6-9e18415b957b.png)

I tried `gitlab -g [target_glab] project create --name sample-app  --import-url [https://url]`, but there's no option to specify username and password as in UI.

Is it supported?

**Comment by nejch** [comment, marked as answer]
@w7089 it doesn't seem like the GitLab API supports supplying this as extra arguments, so python-gitlab won't be able to do it at the moment. You'd have to request this upstream at GitLab first. See available parameters here https://docs.gitlab.com/ee/api/projects.html#create-project.

However, you could probably get around this by supplying auth via `user:name@host`, e.g. providing your import URL as `https://user:password@example.com/repo.git`. But please check if GitLab sanitizes the credentials after the project is created so you don't leak the password in your `import_url` later on.

Also, this is just a simple repository import. You could import a complete GitLab project using `gitlab project-export` and `gitlab project-import`, though that's a bit more involved as you have to download and upload the files between the two instances.

**Comment by w7089** [comment]
Thanks a lot, i succeeded to import using `user:name@host` `import_url`. Actually, i needed the credentials to stay as I imported it as pull mirror. Or you talked about a diff. place?

**Reply by nejch** [reply]
Ah ok, in that case that's actually the documented way of doing it, yes. You should be ok then. I didn't realize you were setting up pull mirroring.

https://docs.gitlab.com/ee/api/projects.html#configure-pull-mirroring-for-a-project

**Reply by w7089** [reply]
actually, that didn't work :(

`gitlab -g [target_glab] project create --name sample-app  --import-url [https://url] --mirror true`

creates the project, but the mirroring is not set...

```
gitlab -g [target_glab] project get --id [id] | grep mirror
mirror: false
```

Should I raise an issue or i'm doing smth. wrong?

**Reply by w7089** [reply]
moreover, `gitlab project mirror-pull --id [id]` shows an exception stacktrace `gitlab.exceptions.GitlabCreateError: 400: The project is not mirrored` instead of a clear message.

**Reply by w7089** [reply]
`gitlab -g [tgt-glab]  project-remote-mirror create --project-id [id]  --url [url] --enabled true` created the mirror, but only push one, didn't see an option to specify pull mirror :(

**Reply by w7089** [reply]
probably GitLab api has a [bug](https://gitlab.com/gitlab-org/gitlab/-/issues/364392) :(. Tried to create the project with pull mirroring enabled using `curl` directly using the API and the mirror wasn't created... only the project was
